File: src/utils/deviceSimulator.py
# deviceSimulator.py
import random

def generate_vital_signs():
    # Randomly select from low, normal, or high ranges
    heart_rate = random.choice([
        random.randint(40, 59),    # Low
        random.randint(60, 100),   # Normal
        random.randint(101, 180)   # High
    ])

    blood_pressure = random.choice([
        f"{random.randint(80, 109)}/{random.randint(50, 69)}",    # Low
        f"{random.randint(110, 140)}/{random.randint(70, 90)}",   # Normal
        f"{random.randint(141, 200)}/{random.randint(91, 130)}"   # High
    ])

    temperature = random.choice([
        round(random.uniform(95.0, 97.4), 1),   # Low
        round(random.uniform(97.5, 99.5), 1),   # Normal
        round(random.uniform(99.6, 105.0), 1)   # High
    ])

    return {
        'heart_rate': heart_rate,
        'blood_pressure': blood_pressure,
        'temperature': temperature
    }

File: src/services/deviceSimulator.js
// deviceSimulator.js
const WebSocket = require('ws');

class DeviceSimulator {
    constructor(wss) {  // Change parameter to accept WebSocket server instance
        this.wss = wss; // Use the passed WebSocket server instance
        this.startStream();
    }

    generateVitalSigns() {
        // Helper function to randomly choose from ranges
        const chooseRange = (ranges) => ranges[Math.floor(Math.random() * ranges.length)];

        const heartRateRanges = [
            () => Math.floor(Math.random() * (59 - 40) + 40),    // Low
            () => Math.floor(Math.random() * (100 - 60) + 60),   // Normal
            () => Math.floor(Math.random() * (180 - 101) + 101)  // High
        ];

        const bpRanges = [
            () => `${Math.floor(Math.random() * (109 - 80) + 80)}/${Math.floor(Math.random() * (69 - 50) + 50)}`,     // Low
            () => `${Math.floor(Math.random() * (140 - 110) + 110)}/${Math.floor(Math.random() * (90 - 70) + 70)}`,   // Normal
            () => `${Math.floor(Math.random() * (200 - 141) + 141)}/${Math.floor(Math.random() * (130 - 91) + 91)}`   // High
        ];

        const tempRanges = [
            () => Number((Math.random() * (97.4 - 95.0) + 95.0).toFixed(1)),    // Low
            () => Number((Math.random() * (99.5 - 97.5) + 97.5).toFixed(1)),    // Normal
            () => Number((Math.random() * (105.0 - 99.6) + 99.6).toFixed(1))    // High
        ];

        return {
            heart_rate: chooseRange(heartRateRanges)(),
            blood_pressure: chooseRange(bpRanges)(),
            temperature: chooseRange(tempRanges)()
        };
    }

    startStream() {
        this.wss.on('connection', (ws) => {
            const interval = setInterval(() => {
                const vitalSigns = this.generateVitalSigns();
                ws.send(JSON.stringify(vitalSigns));
            }, 5000);

            ws.on('close', () => clearInterval(interval));
        });
    }
}

module.exports = DeviceSimulator;
